### ISSUE TITLE
HDF5 1.10.0 breaks NetCDF

### ISSUE THREAD

**Issue by citibeth** [LLNL/spack#856]
@eschnett 

HDF5 was recently updated to version 1.10.0:

https://github.com/citibeth/spack/commit/e436f84244278f842822051a4049ef8418507709#diff-1fbb353310d1439b56a66d8df60ea0f1

Unfortunately, this breaks NetCDF without a corresponding update, see:

http://comments.gmane.org/gmane.comp.lib.netcdf/4073

> The first release candidate for the netCDF-C library version 4.4.1 is now available. There have been a number of improvements since the 4.4.0 release. Most importantly, this release candidate is able to write backwards-compatible netCDF4 files when using HDF5 version 1.10.0. As announced previously, using HDF5 1.10.0 with netCDF 4.4.0 would result in netCDF4 files which were unreadable by systems using an earlier version of libhdf5. See GitHub #250 for more information.

Would it be OK to temporarily comment out version hdf5@1.10.0 until the NetCDF issue is sorted out?

**Comment by tgamblin** [comment]
Can you put `preferred=True` on the prior version?

**Comment by citibeth** [comment]
OK, and here's what I just posted to the list.

The following version constraints have come up with NetCDF:

       netcdf@:4.4.0 requires hdf5@:1.10.0
       If hdf5@1.10.0: is used, then netcdf@4.4.1: is required

QUESTION: How can this constraint best be encoded in netcdf/package.py?  The following is there currently, and needs to not break:

    depends_on("hdf5+mpi~cxx", when='+mpi')  # required for NetCDF-4 support
    depends_on("hdf5~mpi", when='~mpi')  # required for NetCDF-4 support



The first release candidate for the netCDF-C library version 4.4.1 is now available. There have been a number of improvements since the 4.4.0 release. Most importantly, this release candidate is able to write backwards-compatible netCDF4 files when using HDF5 version 1.10.0. As announced previously, using HDF5 1.10.0 with netCDF 4.4.0 would result in netCDF4 files which were unreadable by systems using an earlier version of libhdf5. See GitHub #250 for more information.

**Comment by citibeth** [comment]
@tgamblin  wrote:

```
Elizabeth:

I spoke too soon on github by recommending preferred=True in HDF5.  A better solution is adding this to the netcdf package:

	depends_on('hdf5@:1.9', when='@:4.4.0')
	depends_on('hdf5@1.10:', when='@4.4.1:')

You should be able to put this constraint right below the existing depends_on lines for hdf5.

All the constraints are stored and evaluated one by one, so you actually don't have to bother encoding the constraint above inline with the existing rather complicated constraints.  They'll all be effectively merged during concretization.  Can you see if that will work without the preferred=True?
```
This unfortunately does not work.  See here **without the fix**:
```
[me@ankeli ~]$ spack spec netcdf@4.4.0

Concretized
------------------------------
  netcdf@4.4.0%gcc@4.9.3~hdf4+mpi=linux-x86_64
      ^curl@7.47.1%gcc@4.9.3=linux-x86_64
          ^openssl@system%gcc@4.9.3=linux-x86_64
          ^zlib@1.2.8%gcc@4.9.3=linux-x86_64
      ^hdf5@1.8.16%gcc@4.9.3~cxx~debug+fortran+mpi+shared~szip~threadsafe=linux-x86_64
          ^openmpi@1.10.2%gcc@4.9.3~psm~tm~verbs=linux-x86_64
              ^hwloc@1.11.2%gcc@4.9.3=linux-x86_64
                  ^libpciaccess@0.13.4%gcc@4.9.3=linux-x86_64
                      ^libtool@2.4.6%gcc@4.9.3=linux-x86_64
```

Now see here **with the fix***:
```
[me@ankeli ~]$ spack spec netcdf@4.4.0
Concretized
------------------------------
==> Warning: This installation depends on an old version of OpenSSL, which may have known security issues. 
==> Warning: Consider updating to the latest version of this package.
==> Warning: More details at http://www.openssl.org
==> Error: Invalid spec: 'hdf5@1.8.16%gcc@4.9.3+cxx~debug+fortran~mpi+shared~szip~threadsafe=linux-x86_64'. Package hdf5 requires variant ~cxx, but spec asked for +cxx
```

**Comment by tgamblin** [comment]
That's a bug in spec resolution.  It looks like the `cxx` default is getting applied too soon.  I can take a look but probably not until the weekend.  Let's stick to `preferred=True` for now

**Comment by eschnett** [comment]
@citibeth Has this now been sorted out? Can we remove the preferred version for HDF5? Do you want to add a line such as
```
depends_on('hdf5@:1.8.999', when='@:4.4.0')
```
to NetCDF?

**Comment by citibeth** [comment]
Nothing has changed on the netcdf end.  Ask todd about specpresolution.

**Comment by eschnett** [comment]
@tgamblin I believe the spec resolution bug has been fixed in the mean time; can you confirm?

**Comment by davydden** [comment]
@citibeth i suppose this issue is resolved and can be closed?

**Comment by citibeth** [comment]
Denis,

I have not had a chance to verify what Erik said above.  Someone needs to
change it to the "proper" way suggested by @tgamblin above, and make sure
it works now.  Also, there's now an official release of NetCDF that's more
compatible with past versions of HDF5, so maybe our version jujitsu is no
longer needed.

For all those reasons, NetCDF probably needs to be revisited again before
this is closed.

**Comment by adamjstewart** [comment]
I believe this issue can be closed now that #1291 and #1292 have been merged.
